File: infra/constructs/site.py
import aws_cdk.aws_cloudfront as cloudfront
import aws_cdk.aws_cloudfront_origins as origins
import aws_cdk.aws_route53 as route53
import aws_cdk.aws_route53_targets as route53_targets
import aws_cdk.aws_s3_deployment as s3deploy
from aws_cdk import RemovalPolicy
from aws_cdk.aws_certificatemanager import ICertificate
from aws_cdk.aws_route53 import IHostedZone
from aws_cdk.aws_s3 import Bucket, BucketEncryption
from constructs import Construct


class StaticSite(Construct):
    """S3 bucket, CloudFront distribution, bucket deployment, and DNS A records."""

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        *,
        domain_name: str,
        hosted_zone: IHostedZone,
        certificate: ICertificate,
    ) -> None:
        super().__init__(scope, construct_id)

        bucket = Bucket(
            self,
            "SiteBucket",
            bucket_name=domain_name,
            enforce_ssl=True,
            encryption=BucketEncryption.S3_MANAGED,
            removal_policy=RemovalPolicy.DESTROY,
            auto_delete_objects=True,
        )

        # CloudFront Function to rewrite URLs for clean paths
        # e.g. /resume → /resume/index.html
        url_rewrite_fn = cloudfront.Function(
            self,
            "UrlRewriteFunction",
            code=cloudfront.FunctionCode.from_inline(
                "function handler(event) {"
                "  var request = event.request;"
                "  var uri = request.uri;"
                "  if (uri.endsWith('/')) {"
                "    request.uri += 'index.html';"
                "  } else if (!uri.includes('.')) {"
                "    request.uri += '/index.html';"
                "  }"
                "  return request;"
                "}"
            ),
            runtime=cloudfront.FunctionRuntime.JS_2_0,
        )

        self.distribution = cloudfront.Distribution(
            self,
            "Distribution",
            default_root_object="index.html",
            minimum_protocol_version=cloudfront.SecurityPolicyProtocol.TLS_V1_2_2021,
            default_behavior=cloudfront.BehaviorOptions(
                origin=origins.S3BucketOrigin.with_origin_access_control(bucket),
                viewer_protocol_policy=cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
                allowed_methods=cloudfront.AllowedMethods.ALLOW_GET_HEAD_OPTIONS,
                cache_policy=cloudfront.CachePolicy.CACHING_OPTIMIZED,
                function_associations=[
                    cloudfront.FunctionAssociation(
                        function=url_rewrite_fn,
                        event_type=cloudfront.FunctionEventType.VIEWER_REQUEST,
                    ),
                ],
            ),
            domain_names=[domain_name, f"www.{domain_name}"],
            certificate=certificate,
            error_responses=[
                cloudfront.ErrorResponse(
                    http_status=404,
                    response_http_status=200,
                    response_page_path="/index.html",
                ),
                cloudfront.ErrorResponse(
                    http_status=403,
                    response_http_status=200,
                    response_page_path="/index.html",
                ),
            ],
        )

        s3deploy.BucketDeployment(
            self,
            "DeployWithInvalidation",
            sources=[s3deploy.Source.asset("website/dist")],
            destination_bucket=bucket,
            distribution=self.distribution,
            distribution_paths=["/*"],
        )

        # DNS A records for apex and www
        route53.ARecord(
            self,
            "ApexRecord",
            target=route53.RecordTarget.from_alias(
                route53_targets.CloudFrontTarget(self.distribution),
            ),
            zone=hosted_zone,
            record_name=domain_name,
        )

        route53.ARecord(
            self,
            "WwwRecord",
            target=route53.RecordTarget.from_alias(
                route53_targets.CloudFrontTarget(self.distribution),
            ),
            zone=hosted_zone,
            record_name=f"www.{domain_name}",
        )

File: lib/constructs/site.ts
import { RemovalPolicy } from "aws-cdk-lib";
import type * as acm from "aws-cdk-lib/aws-certificatemanager";
import * as cloudfront from "aws-cdk-lib/aws-cloudfront";
import * as origins from "aws-cdk-lib/aws-cloudfront-origins";
import * as route53 from "aws-cdk-lib/aws-route53";
import * as route53Targets from "aws-cdk-lib/aws-route53-targets";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as s3deploy from "aws-cdk-lib/aws-s3-deployment";
import { Construct } from "constructs";

export interface StaticSiteProps {
	domainName: string;
	hostedZone: route53.IHostedZone;
	certificate: acm.ICertificate;
}

export class StaticSite extends Construct {
	public readonly distribution: cloudfront.Distribution;

	constructor(scope: Construct, id: string, props: StaticSiteProps) {
		super(scope, id);

		const { domainName, hostedZone, certificate } = props;

		const bucket = new s3.Bucket(this, "SiteBucket", {
			bucketName: domainName,
			enforceSSL: true,
			encryption: s3.BucketEncryption.S3_MANAGED,
			removalPolicy: RemovalPolicy.DESTROY,
			autoDeleteObjects: true,
		});

		const urlRewriteFn = new cloudfront.Function(this, "UrlRewriteFunction", {
			code: cloudfront.FunctionCode.fromInline(`
        function handler(event) {
          var request = event.request;
          var uri = request.uri;
          if (uri.endsWith('/')) {
            request.uri += 'index.html';
          } else if (!uri.includes('.')) {
            request.uri += '/index.html';
          }
          return request;
        }
      `),
			runtime: cloudfront.FunctionRuntime.JS_2_0,
		});

		this.distribution = new cloudfront.Distribution(this, "Distribution", {
			defaultRootObject: "index.html",
			minimumProtocolVersion: cloudfront.SecurityPolicyProtocol.TLS_V1_2_2021,
			priceClass: cloudfront.PriceClass.PRICE_CLASS_100,
			defaultBehavior: {
				origin: origins.S3BucketOrigin.withOriginAccessControl(bucket),
				viewerProtocolPolicy: cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
				allowedMethods: cloudfront.AllowedMethods.ALLOW_GET_HEAD_OPTIONS,
				cachePolicy: cloudfront.CachePolicy.CACHING_OPTIMIZED,
				functionAssociations: [
					{
						function: urlRewriteFn,
						eventType: cloudfront.FunctionEventType.VIEWER_REQUEST,
					},
				],
			},
			domainNames: [domainName, `www.${domainName}`],
			certificate,
			errorResponses: [
				{
					httpStatus: 404,
					responseHttpStatus: 200,
					responsePagePath: "/index.html",
				},
				{
					httpStatus: 403,
					responseHttpStatus: 200,
					responsePagePath: "/index.html",
				},
			],
		});

		new s3deploy.BucketDeployment(this, "DeployWithInvalidation", {
			sources: [s3deploy.Source.asset("website/dist")],
			destinationBucket: bucket,
			distribution: this.distribution,
			distributionPaths: ["/*"],
		});

		new route53.ARecord(this, "ApexRecord", {
			target: route53.RecordTarget.fromAlias(
				new route53Targets.CloudFrontTarget(this.distribution),
			),
			zone: hostedZone,
			recordName: domainName,
		});

		new route53.ARecord(this, "WwwRecord", {
			target: route53.RecordTarget.fromAlias(
				new route53Targets.CloudFrontTarget(this.distribution),
			),
			zone: hostedZone,
			recordName: `www.${domainName}`,
		});
	}
}
